### DISCUSSION TITLE
Why can't I do X in Starlark?

### DISCUSSION THREAD

**Discussion by guisellaDes** [kurtosis-tech/kurtosis#1270] [Q&A] [answered]
Starlark seems easy to grok yet more capable than `yaml`. What are its limitations and what should I keep in mind when using Starlark?

**Comment by leeederek** [comment, marked as answer]
Hi @guisellaDes:
Starlark is intended to be a configuration and orchestration language, not a general-purpose programming language. It is excellent at simplicity, readability, and determinism, and terrible at general-purpose programming. We want to use Starlark for what it's good at, while making it easy for you to call down to whatever general-purpose programming you need for more complex logic.

Therefore, Kurtosis provides:

1. [plan.run_sh](https://docs.kurtosis.com/starlark-reference/plan#run_sh) for running Bash tasks on a disposable container
1. [plan.run_python](https://docs.kurtosis.com/starlark-reference/plan#run_python) for running Python tasks on a disposable container
1. [plan.exec](https://docs.kurtosis.com/starlark-reference/plan#exec) for running Bash on a service

All of these let you customize the image to run on, so you can functionally call any code in any language using Kurtosis.

To learn more about why we chose Starlark and how it gets implemented at Kurtosis, we encourage you to check out our docs here on this topic: https://docs.kurtosis.com/concepts-reference/starlark
